### DISCUSSION TITLE
Could not load file or assembly "CefSharp.core.runtime.dll" when using VCRuntime.CefSharp.140 Nuget package.

### DISCUSSION THREAD

**Discussion by uptodatehuang** [cefsharp/CefSharp#3422] [General]
**What version of the product are you using?**
NuGet:
cef.redist.x64(88),
cef.redist.x86(88)
CefSharp.Common(88-pre-release)
CefSharp.WinForms(88-pre-release)
VCRuntime.CefSharp.140(1.0.4)

**What architecture x86 or x64?**
x86

**On what operating system?**
Win10

**Are you using WinForms, WPF or OffScreen?**
WinForms

**What steps will reproduce the problem?**
1. tested app working with Visual C++ Redistributables 2015 installed in PC.
2. then remove any Visual C++ Redistributables from the PC using 'Add or Remove Program' feature
3.copy below files into application folder
concrt140.dll
msvcp140.dll
msvcp140_1.dll
msvcp140_2.dll
vccorlib140.dll
vcruntime140.dll

**What is the expected output? What do you see instead?**
Expected: Cef should initiate properly
Actual: throw 'Unexpected Exception' during Cef Ini and error message is:  
CefSharp could not load file or assembly "CefSharp.core.runtime.dll" or one of its dependencies. is not a valid Win32 application

**Please provide any additional information below.**

**Reference:**
https://stackoverflow.com/questions/54755445/how-to-include-vc-redistributable-files-needed-by-cefsharp

**Comment by uptodatehuang** [comment]
We found the issue.

We try to use this Nuget package to bring in VC++ runtime files: **VCRuntime.CefSharp.140**
But it actually just bring in **msvcp140.dll and vcruntime140.dll** into application folder.
So we manually copied the rest of files into application folder and keep these 2 files untouched.
concrt140.dll
msvcp140_1.dll
msvcp140_2.dll
vccorlib140.dll
Then we got the error.

The fix we did is: 
- uninstall this Nuget package **VCRuntime.CefSharp.140**
- copy all files inside this folder C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\VC\Redist\MSVC\14.15.26706\x86\Microsoft.VC141.CRT to the application folder, and the app starts properly.

Summary:
I think the root cause is incompartibility between the files this package bring and the rest of files from my pc

**Reply by amaitland** [reply]
> VCRuntime.CefSharp.140(1.0.4)

Whilst the package has CefSharp in the name it has no association with the CefSharp project. I've never used it.

https://github.com/cefsharp/CefSharp/wiki/Frequently-asked-questions#6-how-do-i-include-the-visual-studio-c-201220132015-redistributables-with-my-application has details on including vc++ with your application. It's a little outdated, anyone with a GitHub account can edit the wiki, your welcome to make improvements.

**Reply by amaitland** [reply]
If you are looking at automating the approach then I'd suggest using something like the following:

```xml
  <Import Project="$(DevEnvDir)..\..\VC\Auxiliary\Build\Microsoft.VCToolsVersion.default.props"/>
  
  <Target Name="IncludeVisualCppInOutFolder" AfterTargets="ResolveReferences">
    <PropertyGroup>
      <!-- For VS2017 we use VC141, for VS2019 we use VC142 -->
      <_VCVersion>VC141</_VCVersion>
      <_VCVersion Condition="'$(VisualStudioVersion)'=='16.0'">VC142</_VCVersion>
      <_VCRedistLocation>$(DevEnvDir)..\..\VC\Redist\MSVC\$(VCToolsRedistVersion)\$(PlatformTarget)\Microsoft.$(_VCVersion).CRT</_VCRedistLocation>
    </PropertyGroup>
    <Message Importance="high" Text="_VCVersion = $(_VCVersion)" />
    <Message Importance="high" Text="_VCRedistLocation = $(_VCRedistLocation)" />
    <ItemGroup>
      <ReferenceCopyLocalPaths  Include="$(_VCRedistLocation)\**\*" />
    </ItemGroup>
  </Target>
```

- Only works when run under `Visual Studio`
- Only works for `x86/x64` platformtargets
- Should work under `VS2017` though not tested.
